File: 다익스트라/1238번 파티.cpp
#include<iostream>
#include<vector>
#include<queue>

#define MAX 1001
#define INF 987654321
#define P pair<int, int>

using namespace std;

int n, m, x, ans = 0;
int temp[MAX], result[MAX];
vector<P> vec[MAX];
priority_queue<P, vector<P>, greater<P>> pque;

void input()
{
	cin >> n >> m >> x;

	for (int i = 0; i < m; ++i) {
		int from, to, time;
		cin >> from >> to >> time;

		vec[from].push_back({ to, time });
	}
}

void Dijkstra(int start)
{
	temp[start] = 0;
	pque.push({ 0, start });

	while (!pque.empty()) {
		int time = pque.top().first;
		int node = pque.top().second;
		pque.pop();

		for (int i = 0; i < vec[node].size(); ++i) {
			int next_node = vec[node][i].first;
			int next_time = vec[node][i].second;

			if (temp[next_node] > next_time + time) {
				temp[next_node] = next_time + time;
				pque.push({ temp[next_node], next_node });
			}
		}
	}
}

void solve()
{
	for (int i = 1; i <= n; ++i) {
		for (int j = 1; j <= n; ++j)
			temp[j] = INF;
		Dijkstra(i);
		result[i] = temp[x];
	}

	for (int i = 1; i <= n; ++i)
		temp[i] = INF;
	Dijkstra(x);

	for (int i = 1; i <= n; ++i) {
		result[i] += temp[i];
		ans = max(ans, result[i]);
	}
		
	cout << ans;
}

void init()
{
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);
}

int main()
{
	init();
	input();
	solve();

	return 0;
}

File: 다익스트라/1238번 파티.py
import sys
import heapq

# 전처리 부분
input = sys.stdin.readline
INF = sys.maxsize

# 알고리즘 부분
def dijkstra(start):
    heap = [[0, start]]
    time[start] = 0

    while heap:
        cost, node = heapq.heappop(heap)

        for n_cost, n_node in edges[node]:
            if time[n_node] > cost + n_cost:
                time[n_node] = cost + n_cost
                heapq.heappush(heap, [time[n_node], n_node])

# 변수 선언 및 초기화 부분
n, m, x = map(int, input().split())
edges = [[] for _ in range(n + 1)]
for _ in range(m):
    a, b, c = map(int, input().split())
    edges[a].append([c, b])
result = [0] * (n + 1)
ans = 0

# 메인 코드 부분
for i in range(1, n + 1):
    time = [INF] * (n + 1)
    dijkstra(i)
    result[i] = time[x]

time = [INF] * (n + 1)
dijkstra(x)

for i in range(1, n + 1):
    result[i] += time[i]
    ans = max(ans, result[i])

print(ans)